src/NUnitFramework/framework/Internal/Filters/NotFilter.cs

```diff
@@ -54,13 +54,11 @@ public NotFilter( TestFilter baseFilter)
         /// any child test matches.
         /// </summary>
         /// <param name="test">The test to which the filter is applied</param>
+        /// <param name="negated">If set to <c>true</c> we are carrying a negation through</param>
         /// <returns>True if the test passes the filter, otherwise false</returns>
-        public override bool Pass(ITest test)
+        public override bool Pass(ITest test, bool negated = false)
```

Comment:
This is very clever and obviously central to the algorithm. Others should decide whether this is a breaking change. Obviously, it changes the interface by changing a method, rather than merely adding one. Are there .NET languages for which the use of a default value is non-compliant?

Reply:
No problem @CharliePoole :)

unfortunately, I don't know about the compliance, but I will be happy to change it to a separate method if this is preferable. This was actually my thought to begin with, until I stated making the change. I went with this solution because I thought it might be more conceptually clear. However, this is probably not much, and might even entirely be my subjective view.

Reply:
If there is an already-compiled assembly out there that calls one of these `Pass` methods, this change will result in MissingMethodException because the compiled binary has a hard reference to the `bool Pass(ITest)` signature and the .NET runtime doesn't have the ability to forward that to the new `bool Pass(ITest, bool)` signature.

Adding an interface method is also a breaking change for anyone out there who implements ITestFilter themselves. Once published, interfaces can never be changed without breaking somebody: https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md#members. This was the motivation for the C# 8 Default Interface Members feature which only works on the .NET Core 3.0+ runtime, but that's not useful to us unless we stop supporting .NET Framework.

Can we think up a way around this? The .NET Framework used the tactic of adding virtual methods to abstract base classes implementing the existing functions, such as `DbConnection`, since `IDbConnection` can never be changed.

We've had several complaints about changes like this, so I want to either do something non-breaking or wait until NUnit 4 to fix the bug.

Reply:
If there is an already-compiled assembly out there that calls one of these `Pass` methods, this change will result in MissingMethodException because the compiled binary has a hard reference to the `bool Pass(ITest)` signature and the .NET runtime doesn't have the ability to forward that to the new `bool Pass(ITest, bool)` signature.

Adding an interface method is also a breaking change for anyone out there who implements ITestFilter themselves. Once published, interfaces can never be changed without breaking somebody: https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md#members. This was the motivation for the C# 8 Default Interface Members feature which only works on the .NET Core 3.0+ runtime, but that's not useful to us unless we stop supporting .NET Framework.

Can we think up a way around this? The .NET Framework used the tactic of adding virtual methods to abstract base classes implementing the existing functions, such as `DbConnection`, since `IDbConnection` can never be changed.

We've had several complaints about changes like this, so I want to either do something non-breaking or wait until NUnit 4 to fix the bug.

Reply:
@jnm2 In spite of the Microsoft guidelines, I think we could add a member here without much impact.

Of course it is possible that somebody has taken our code and used it in unexpected ways. However...

1. There is no way to create a usable custom filter without modifying the framework. In practice this means that any new filter has to be implemented by us and is peculiar to the version of the framework that contains support for it. That is to say, there is no extensibility mechanism for filters.

2. The filter interface is really internal, as are almost all of those in the `Interfaces` directory. I believe they were once in the `Internal` directory and were moved to reduce clutter - probably not a good decision. It's primary use is to allow easy testing.

3. The API for the framework is reflection-based rather than interface-based anyway.

4. Microsoft's rules are based on their bad experience. Someone in the company has said that anything you make accessible to users immediately becomes a feature that must be supported. We can all sympathize with that! However, I don't think we need to support unreasonable usage of accessible members, i.e. usage that goes beyond the fairly obvious intention of the interface, which in this case is to get the framework to filter tests.

5. This __will__ impact anyone who has modified the source of the framework for their own purposes, adding new filters. But only if they decide to make the same change themselves, which seems fine to me.

`<meta>`
It would be great if we could have a written, approved and published standard around breaking changes. We've been pushing around the same arguments for a couple of years now without coming to much conclusion. The key thing to define IMO is what is meant by a "published API".
`</meta>`

Reply:
Okay, if we are sure that no one will be implementing it, then adding a new method will be fine. I haven't tried to look for a problem with that, but I trust you.

> The API for the framework is reflection-based rather than interface-based anyway.

That's only the API for driving the framework from an engine. The Assert class and everything users write `using NUnit.Framework` for are another API, one which we care about because users reasonably expect that we will not break their libraries or code (unless maybe under Internal, but even there we tell them that is their only option sometimes and we still shouldn't break that).

Reply:
As I've written before we have a number of APIs. I believe each of them has to have its own standard. Here, I was talking about the "Framework API" for calling the framework to run tests. The name comes from our docs.

Asserts, etc. are part of the test-writing API, which we don't call out as such, both obviously exists. There is also an extensibility API, which is partially documented.

Reply:
Right, so then I'm not sure how point 3 has any bearing on this PR and the complaints I mentioned that we've been getting when we change APIs that we ship out of this repo.

Reply:
Only bearing was if you were considering this as part of that API.

Reply:
Okay, no. Since few people maintain NUnit engines/drivers, most of my thought goes to the user-testing and extensibility APIs that we get the complaints about.